### PR TITLE
Feature/week 3

### PR DESCRIPTION
**Pull request by Marty-Hofstra**
## Description
This PR contains changes that add a Feature Serving task to the repo, this is done by adding a Serving class. Moreover, a featurisation class has been created, replacing the existing feature util function. Unfortunately, I could not perform Model Serving due to the PySpark dependency not being available for online serving. Hence, I've only done Feature Serving in this assignment.

## Release Notes

- **New Features**
  - Introduced a `Serving` class for managing prediction serving endpoints in Databricks.
  - Added a `Featurisation` class to encapsulate feature table writing functionality.
  - Implemented a new workflow for serving features in a hotel reservation context.
  
- **Documentation**
  - Updated `README.md` with clearer usage instructions and examples for the `hotel_reservations` package.
  - Enhanced `CHANGELOG.md` to reflect the latest version updates and new features.

- **Tests**
  - Added test suites for both `Featurisation` and `Serving` classes to ensure functionality and error handling.
  
- **Chores**
  - Updated `.gitignore` to exclude the `reports/` directory from version control.
  - Incremented project version in `pyproject.toml` to `0.1.0`.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,15 +4,28 @@ All notable changes to the `hotel_reservations` project will be documented in th
 The format is based on [Keep a Changelog](http://keepachangelog.com/)
 and this project adheres to [Semantic Versioning](http://semver.org/).
 
+## [0.0.3] - 2024-11-10
+
```
[the remark this answers is not in the document]
Comment (Marty-Hofstra) [reply CHANGELOG.md]: Good suggestion!

File: test/serving/test_serving.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        "https://test_host/serving-endpoints/test_endpoint/invocations",
+        headers={"Authorization": "Bearer test_token"},
+        json={"dataframe_records": [{"test_pk": pk_value}]},
+    )
+
+    # Check response and latency
+    assert response_status == 200
+    assert response_text == "success"
+    assert latency == 1  # Since we mocked time.time to return 1 and 2, latency = 2 - 1 = 1
+
```
[the remark this answers is not in the document]
Comment (Marty-Hofstra) [reply test/serving/test_serving.py]: Docstrings for unit tests seems overkill, but I've added a unit test on the error cases and one on a timeout

File: test/serving/test_serving.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    response_status, response_text, latency = serving_instance.send_request_random_id(id_list)
+
+    # Verify that random.choice was called and send_request was called with the selected id
+    mock_random_choice.assert_called_once_with(id_list)
+    mock_send_request.assert_called_once_with("random_id")
+
+    # Check response and latency
+    assert response_status == 200
+    assert response_text == "success"
+    assert latency == 1.5
```
[the remark this answers is not in the document]
Comment (Marty-Hofstra) [reply test/serving/test_serving.py]: Good one, also added a check in the function itself that checks if the list is empty, and if not, raises a ValueError

File: README.md [hunk 4]
```diff
@@ -31,8 +31,11 @@ The `hotel_reservations` package (as a .whl) can be created and stored in DBFS b
 In the cluster configuration, click on `Libraries` and then `Install new`, select `Volumes` and navigate to the path where you stored the wheel. The package functionality can be imported with `import hotel_reservations`. An example of usage of the data processing functions is as follows:
 
 ```
-with open("../../../project_config.yaml", "r") as file:
-    config = yaml.safe_load(file)
+from pyspark.sql import SparkSession
+from hotel_reservations.data_processing.data_processor import DataProcessor
+from hotel_reservations.utils import open_config
+
+config = open_config("../../../../project_config.yaml", scope="marty-MLOPs-cohort")
```
[the remark this answers is not in the document]
Comment (Marty-Hofstra) [reply README.md]: It is expected that users of this repo will do so through DBConnect or in Databricks (by syncing the workspace). Hence, Databricks specific functionality can stay here. However, demonstrating some Feature Serving functionality would be a good idea

File: test/featurisation/test_featurisation.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    result = featurisation_instance.enable_change_data_feed(spark)
+
+    # Check if the SQL query called contains the correct statements without exact whitespace matching
+    args, _ = mock_sql.call_args
+    assert "ALTER TABLE" in args[0]
+    assert featurisation_instance.feature_table_name in args[0]
+    assert "SET TBLPROPERTIES (delta.enableChangeDataFeed = true)" in args[0]
+
+    assert result == f"Change data feed has been enabled for {featurisation_instance.feature_table_name}."
+
```
[the remark this answers is not in the document]
Comment (Marty-Hofstra) [reply test/featurisation/test_featurisation.py:68]: This doesn't work due to differences in whitespaces

File: src/hotel_reservations/serving/serving.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        """Sends a request to the endpoint with the primary key value as input
+
+        Args:
+            pk_value (str): Value of the primary key for which the prediction is requested
+
+        Returns:
+            reponse_status (str): Status code of the response
+            reponse_text (str): The content of the response
+            latency (float): Latency in seconds
+        """
```
[the remark this answers is not in the document]
Comment (Marty-Hofstra) [reply src/hotel_reservations/serving/serving.py]: Sharp, thanks!

File: src/hotel_reservations/serving/serving.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        total_start_time = time.time()
+        latencies: list[float] = []
+
+        with ThreadPoolExecutor(max_workers=max_workers) as executor:
+            futures = [executor.submit(self.send_request_random_id, pk_list) for _ in range(self.num_requests)]
+
+            for future in as_completed(futures):
+                latency = future.result()[2]
+                latencies.append(latency)
```
[the remark this answers is not in the document]
Comment (Marty-Hofstra) [reply src/hotel_reservations/serving/serving.py]: Added

File: src/hotel_reservations/serving/serving.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        start_time = time.time()
+        serving_endpoint = f"https://{self.host}/serving-endpoints/{self.serving_endpoint_name}/invocations"
+        response = requests.post(
+            f"{serving_endpoint}",
+            headers={"Authorization": f"Bearer {self.token}"},
+            json={"dataframe_records": [{self.primary_key: pk_value}]},
+        )
+        end_time = time.time()
+        latency = end_time - start_time
+
```
[the remark this answers is not in the document]
Comment (Marty-Hofstra) [reply src/hotel_reservations/serving/serving.py]: I think the current try/except clause is sufficient

File: src/hotel_reservations/serving/serving.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+                            entity_name=feature_spec_name,
+                            scale_to_zero_enabled=True,
+                            workload_size=workload_size,
+                        )
+                    ]
+                ),
+            )
+        except Exception as e:
+            print(f"Serving endpoint '{self.serving_endpoint_name}' not created: {e}")
+
```
[the remark this answers is not in the document]
Comment (Marty-Hofstra) [reply src/hotel_reservations/serving/serving.py]: nice one, thanks!

File: README.md [hunk 10]
```diff
@@ -49,3 +58,67 @@ Y_test = test.select(config.target)
...
+```
+from pyspark.sql import SparkSession
+
+from hotel_reservations.featurisation.featurisation import Featurisation
+from hotel_reservations.utils import open_config
+
+spark = SparkSession.builder.getOrCreate()
+
+# Load config from a relative path and add your Databricks secret scope here
+config = open_config("../../../../project_config.yaml", scope="databricks_secret_scope")
```
[the remark this answers is not in the document]
Comment (Marty-Hofstra) [reply README.md]: Changed

**Review by basakeskili — APPROVED** [review]

**Pull request merged**